### PR TITLE
feat(cli): adds `--dump-state` cli cmd 

### PR DESCRIPTION
**Pull request by dutterbutter**
# What :computer: 
* adds `--dump-state` cli cmd 
* adds `--state-interval` cli cmd
* Partial implementation of https://github.com/matter-labs/anvil-zksync/issues/482
* Updates alloy-zksync in e2e

# Why :hand:
* anvil parity

# Evidence :camera:
Include screenshots, screen recordings, or `console` output here demonstrating that your changes work as intended

### Usage

Run the following command, wait ~60 secs and view `state.json` file:
```
./target/release/anvil-zksync --dump-state state.json run
```

Forking:
```
./target/release/anvil-zksync --dump-state state.json fork --fork-url sepolia-testnet
```

### CODE REVIEW

File: crates/cli/src/main.rs
Comment (dutterbutter) [inline crates/cli/src/main.rs:367]: @itegulov inspecting the state dump we always get:

`"l1BatchNumber": null,` 

This is problematic when you attempt to load that same dumped state:
```
thread 'main' panicked at crates/core/src/node/in_memory.rs:925:13:
encountered a block with no batch; this is not supposed to happen
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Comment (itegulov) [reply crates/cli/src/main.rs:367]: Very strange, this shouldn't happen indeed. Do you have an example of how I can reproduce this behaviour?

File: crates/cli/src/main.rs
Comment (itegulov) [reply crates/cli/src/main.rs:346]: Actually, looking at the code it seems like `PeriodicStateDumper` is designed to work even if `dump_path` is not present. So I would remove the whole `if` here.

File: crates/cli/src/main.rs
Comment (itegulov) [inline crates/cli/src/main.rs:369]: If this is going to be our "exit point" then could you move `BlockProducer` here as well? So it's a part of `select!` and is not spawned on its own.

**Comment by itegulov** [comment]
Also, would love to see an e2e test for this

File: crates/cli/src/main.rs
Comment (dutterbutter) [reply crates/cli/src/main.rs:347]: Yup, nice catch. Done ✅

File: crates/core/src/node/mod.rs
Comment (dutterbutter) [reply crates/core/src/node/mod.rs]: Done ✅

File: crates/cli/src/main.rs
Comment (dutterbutter) [reply crates/cli/src/main.rs]: Yes, to reproduce `./target/release/anvil-zksync --state-interval 1 --dump-state state.json fork --fork-url mainnet` 

Seems specific to forking?

**Comment by dutterbutter** [comment]
@itegulov added e2e tests for `run` and `fork` for `--dump_state`. I had to update the alloy-zksync dep, hence the extra changes.

File: crates/cli/Cargo.toml
Comment (itegulov) [reply crates/cli/Cargo.toml:39]: Ah, I forgot to add core's dev-dependencies to root workspace. They should be there as should this :sweat_smile:

File: crates/cli/src/main.rs
Comment (itegulov) [reply crates/cli/src/main.rs]: Ahhh, yes totally possible with forking. I didn't realize we save fork's latest block to local storage on initialization. Do you know if that's necessary to do at all? It kind of breaks a few invariants in our code.

Comment (itegulov) [reply crates/cli/src/main.rs]: In any case feel free to remove that line and see if that resolves the issue for you. If not I can take another look

File: crates/cli/src/main.rs
Comment (itegulov) [reply crates/cli/src/main.rs]: I incorporates this approach in https://github.com/matter-labs/anvil-zksync/pull/497, feel free to rebase on those changes and just add state dumper to the existing `select!` now

File: crates/cli/src/main.rs
Comment (dutterbutter) [reply crates/cli/src/main.rs]: hmm I will look to resolve in https://github.com/matter-labs/anvil-zksync/pull/502

File: e2e-tests-rust/Cargo.toml
Comment (itegulov) [inline e2e-tests-rust/Cargo.toml:13]: :+1:

File: e2e-tests-rust/Cargo.toml
Comment (itegulov) [inline e2e-tests-rust/Cargo.toml:14]: Created an issue for upgrading alloy to match alloy-zksync https://github.com/matter-labs/anvil-zksync/issues/504

(nothing actionable here just noting)

File: e2e-tests-rust/tests/lib.rs
Comment (itegulov) [inline e2e-tests-rust/tests/lib.rs:24]: Please move these above the constants. Also seems like rustfmt+clippy are not enforced here so I created a separate issue https://github.com/matter-labs/anvil-zksync/issues/505

In general I recommend enabling rustfmt-based auto-formatting in your editor (not sure what you use but there should be an option for almost anything)

File: e2e-tests-rust/tests/lib.rs
Comment (itegulov) [inline e2e-tests-rust/tests/lib.rs:537]: I think this wouldn't be necessary if you added `state_dumper.dump()` at the end of the program like I suggested in that other thread, right?

File: e2e-tests-rust/tests/lib.rs
Comment (itegulov) [inline e2e-tests-rust/tests/lib.rs:549]: nit: use `.with_context` for better readability

File: e2e-tests-rust/tests/lib.rs
Comment (itegulov) [inline e2e-tests-rust/tests/lib.rs:602]: nit: same thing here

File: e2e-tests-rust/tests/lib.rs
Comment (itegulov) [inline e2e-tests-rust/tests/lib.rs:558]: I would also assert that it contains that one transaction we executed at the beginning

File: e2e-tests-rust/tests/lib.rs
Comment (itegulov) [inline e2e-tests-rust/tests/lib.rs:611]: Same here

**Review by itegulov — APPROVED** [review]
Alright looks good to me. Happy to merge this if you want, all my suggestions are small and can come in a separate PR

File: e2e-tests-rust/tests/lib.rs
Comment (dutterbutter) [reply e2e-tests-rust/tests/lib.rs:24]: Opened PR for #505 here: https://github.com/matter-labs/anvil-zksync/pull/506

File: e2e-tests-rust/Cargo.toml
Comment (dutterbutter) [reply e2e-tests-rust/Cargo.toml:14]: Will pick #504 once this is merged

**Comment by dutterbutter** [comment]
Will make changes in subsequent PR.

File: Cargo.toml
Comment (dutterbutter) [reply Cargo.toml:84]: Yes, its used in unit tests in core.

Comment (itegulov) [reply Cargo.toml:84]: https://github.com/matter-labs/anvil-zksync/issues/507

**Pull request merged**
